File: ios/App/App/TVRemotePlugin.swift
import Capacitor
import Foundation

@objc(TVRemotePlugin)
public class TVRemotePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "TVRemotePlugin"
    public let jsName = "TVRemote"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getPermissionState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "discoverDevices", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addManualDevice", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pairDevice", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listPairedDevices", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removePairedDevice", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "validateDevice", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "testPlay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "armPlay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelArmedPlay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendPlayNow", returnType: CAPPluginReturnPromise)
    ]

    private let engine = TVRemoteEngine()

    @objc func getPermissionState(_ call: CAPPluginCall) {
        call.resolve(["localNetwork": engine.localNetworkPermissionState()])
    }

    @objc func requestPermissions(_ call: CAPPluginCall) {
        call.resolve(["localNetwork": engine.localNetworkPermissionState()])
    }

    @objc func discoverDevices(_ call: CAPPluginCall) {
        call.resolve(["devices": engine.discoverManualOnly()])
    }

    @objc func addManualDevice(_ call: CAPPluginCall) {
        do {
            let protocolId = call.getString("protocol") ?? "manual"
            let host = call.getString("host") ?? ""
            let port = call.getInt("port")
            let device = try engine.addManualDevice(protocolId: protocolId, host: host, port: port)
            call.resolve(device)
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func pairDevice(_ call: CAPPluginCall) {
        call.resolve(["ok": false, "requiresPin": false, "errorCode": "PAIRING_NOT_IMPLEMENTED", "errorMessage": "Pairing is scaffolded; hardware flow is not enabled yet."])
    }

    @objc func listPairedDevices(_ call: CAPPluginCall) {
        call.resolve(["devices": engine.registry.list()])
    }

    @objc func removePairedDevice(_ call: CAPPluginCall) {
        let deviceId = call.getString("deviceId") ?? ""
        engine.registry.remove(id: deviceId)
        call.resolve(["ok": true])
    }

    @objc func validateDevice(_ call: CAPPluginCall) {
        let deviceId = call.getString("deviceId") ?? ""
        engine.validateDevice(deviceId: deviceId) { result in
            call.resolve(result)
        }
    }

    @objc func testPlay(_ call: CAPPluginCall) {
        let deviceId = call.getString("deviceId") ?? ""
        engine.testPlay(deviceId: deviceId) { result in
            call.resolve(result)
        }
    }

    @objc func armPlay(_ call: CAPPluginCall) {
        let countdownId = call.getString("countdownId") ?? ""
        let deviceId = call.getString("deviceId") ?? ""
        let playAtServerMs = call.getDouble("playAtServerMs") ?? 0
        let playAtMonotonicMs = call.getDouble("playAtMonotonicMs") ?? 0
        engine.armPlay(countdownId: countdownId, deviceId: deviceId, playAtServerMs: playAtServerMs, playAtMonotonicMs: playAtMonotonicMs) { result in
            call.resolve(result)
        }
    }

    @objc func cancelArmedPlay(_ call: CAPPluginCall) {
        let countdownId = call.getString("countdownId") ?? ""
        let deviceId = call.getString("deviceId") ?? ""
        call.resolve(["ok": engine.commandScheduler.cancel(countdownId: countdownId, deviceId: deviceId)])
    }

    @objc func sendPlayNow(_ call: CAPPluginCall) {
        let countdownId = call.getString("countdownId") ?? ""
        let deviceId = call.getString("deviceId") ?? ""
        engine.sendPlayNow(countdownId: countdownId, deviceId: deviceId) { result in
            call.resolve(result)
        }
    }
}

File: android/app/src/main/java/tech/kyrosdirect/play321/TVRemotePlugin.kt
package tech.kyrosdirect.play321

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "TVRemote")
class TVRemotePlugin : Plugin() {
    @PluginMethod fun getPermissionState(call: PluginCall) { call.resolve(JSObject().put("localNetwork", "not_required")) }
    @PluginMethod fun requestPermissions(call: PluginCall) { call.resolve(JSObject().put("localNetwork", "not_required")) }
    @PluginMethod fun discoverDevices(call: PluginCall) { call.resolve(JSObject().put("devices", JSArray().put(manualCandidate()))) }
    @PluginMethod fun addManualDevice(call: PluginCall) {
        val protocol = call.getString("protocol", "manual") ?: "manual"
        val host = call.getString("host", "") ?: ""
        if (protocol != "manual" && host.isBlank()) {
            call.reject("TV host/IP is required.")
            return
        }
        call.resolve(JSObject().put("id", "$protocol:$host").put("protocol", protocol).put("displayName", if (protocol == "manual") "Manual countdown" else "$protocol device ($host)").put("host", host).put("requiresPairing", protocol != "roku-ecp").put("capabilities", JSArray()))
    }
    @PluginMethod fun pairDevice(call: PluginCall) { call.resolve(unavailable("PAIRING_NOT_IMPLEMENTED", "Android pairing scaffold exists but is not implemented yet.")) }
    @PluginMethod fun listPairedDevices(call: PluginCall) { call.resolve(JSObject().put("devices", JSArray())) }
    @PluginMethod fun removePairedDevice(call: PluginCall) { call.resolve(JSObject().put("ok", true)) }
    @PluginMethod fun validateDevice(call: PluginCall) { call.resolve(unavailable("ADAPTER_NOT_IMPLEMENTED", "Android native adapter validation is scaffolded only.")) }
    @PluginMethod fun testPlay(call: PluginCall) { call.resolve(commandUnavailable("ADAPTER_NOT_IMPLEMENTED", "Android native Test Play is scaffolded only.")) }
    @PluginMethod fun armPlay(call: PluginCall) { call.resolve(JSObject().put("ok", false).put("deviceId", call.getString("deviceId", "") ?: "").put("countdownId", call.getString("countdownId", "") ?: "").put("armedForMonotonicMs", call.getDouble("playAtMonotonicMs", 0.0)).put("errorCode", "ADAPTER_NOT_IMPLEMENTED").put("errorMessage", "Android native Auto Play arming is scaffolded only.")) }
    @PluginMethod fun cancelArmedPlay(call: PluginCall) { call.resolve(JSObject().put("ok", true)) }
    @PluginMethod fun sendPlayNow(call: PluginCall) { call.resolve(commandUnavailable("ADAPTER_NOT_IMPLEMENTED", "Android native sendPlayNow is scaffolded only.")) }

    private fun manualCandidate() = JSObject().put("id", "manual-countdown").put("protocol", "manual").put("displayName", "Manual countdown").put("requiresPairing", false).put("capabilities", JSArray())
    private fun unavailable(code: String, message: String) = JSObject().put("ok", false).put("errorCode", code).put("errorMessage", message)
    private fun commandUnavailable(code: String, message: String) = unavailable(code, message).put("protocol", "manual").put("command", "play").put("sentAtMonotonicMs", System.currentTimeMillis().toDouble())
}
